core/src/main/java/com/linecorp/armeria/common/stream/InputStreamStreamMessage.java

```diff
@@ -0,0 +1,380 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common.stream;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static com.linecorp.armeria.internal.common.stream.InternalStreamMessageUtil.containsNotifyCancellation;
+import static java.util.Objects.requireNonNull;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.math.LongMath;
+import com.google.common.primitives.Ints;
+
+import com.linecorp.armeria.common.CommonPools;
+import com.linecorp.armeria.common.HttpData;
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.common.util.EventLoopCheckingFuture;
+import com.linecorp.armeria.internal.common.stream.NoopSubscription;
+import com.linecorp.armeria.server.ServiceRequestContext;
+
+import io.netty.util.concurrent.EventExecutor;
+
+final class InputStreamStreamMessage implements ByteStreamMessage {
+
+    private static final Logger logger = LoggerFactory.getLogger(InputStreamStreamMessage.class);
+
+    private static final AtomicIntegerFieldUpdater<InputStreamStreamMessage> subscribedUpdater =
+            AtomicIntegerFieldUpdater.newUpdater(InputStreamStreamMessage.class, "subscribed");
+
+    private final InputStream inputStream;
+    @Nullable
+    private final ExecutorService blockingTaskExecutor;
+    private final int bufferSize;
+
+    private long offset;
+    private long length = Long.MAX_VALUE;
+
+    private volatile int subscribed;
+    private final CompletableFuture<Void> completionFuture = new EventLoopCheckingFuture<>();
+
+    @Nullable
+    private volatile InputStreamSubscription inputStreamSubscription;
+
+    InputStreamStreamMessage(InputStream inputStream, @Nullable ExecutorService blockingTaskExecutor,
+                             int bufferSize) {
+        requireNonNull(inputStream, "inputStream");
+        this.inputStream = inputStream;
+        this.blockingTaskExecutor = blockingTaskExecutor;
+        this.bufferSize = bufferSize;
+    }
+
+    @Override
+    public ByteStreamMessage range(long offset, long length) {
+        checkArgument(offset >= 0, "offset: %s (expected: >= 0)", offset);
+        checkArgument(length > 0, "length: %s (expected: > 0)", length);
+        checkState(subscribed == 0, "cannot specify range(%s, %s) after this %s is subscribed",
+                   offset, length, InputStreamStreamMessage.class);
```

Comment:
Maybe remove `InputStreamStreamMessage.class` and say `once this stream is subscribed`?

The resulting exception message is currently like `this class InputStreamStreamMessage is subscribed`, which sounds somewhat awkward.

Reply:
✅ updated!